File: Frontend/lib/features/crm/data/lookup_service.dart
import 'dart:convert';

import '../../../core/network/api_client.dart';

const defaultPipelineStages = [
  'QUALIFICATION',
  'TECHNICAL_EVAL',
  'BUDGET_VALIDATION',
  'PROPOSAL',
  'QUOTATION_ISSUED',
  'NEGOTIATION',
];

String? nextPipelineStage(List<String> stages, String current) {
  final idx = stages.indexOf(current);
  if (idx < 0 || idx >= stages.length - 1) return null;
  return stages[idx + 1];
}

const outcomeRequiredTypes = {'CALL', 'MEETING', 'TASK'};

class ActivityType {
  ActivityType({
    required this.activityTypeId,
    required this.code,
    required this.name,
    required this.isActive,
  });

  final String activityTypeId;
  final String code;
  final String name;
  final bool isActive;

  factory ActivityType.fromJson(Map<String, dynamic> json) => ActivityType(
        activityTypeId: json['activity_type_id'] as String,
        code: json['code'] as String,
        name: json['name'] as String,
        isActive: json['is_active'] as bool? ?? true,
      );
}

class ActivityOutcome {
  ActivityOutcome({
    required this.activityOutcomeId,
    required this.activityTypeCode,
    required this.code,
    required this.name,
    required this.isPositive,
    required this.isActive,
  });

  final String activityOutcomeId;
  final String activityTypeCode;
  final String code;
  final String name;
  final bool isPositive;
  final bool isActive;

  factory ActivityOutcome.fromJson(Map<String, dynamic> json) => ActivityOutcome(
        activityOutcomeId: json['activity_outcome_id'] as String,
        activityTypeCode: json['activity_type_code'] as String,
        code: json['code'] as String,
        name: json['name'] as String,
        isPositive: json['is_positive'] as bool? ?? false,
        isActive: json['is_active'] as bool? ?? true,
      );
}

class OpportunityStage {
  OpportunityStage({
    required this.opportunityStageId,
    required this.code,
    required this.name,
    required this.sequenceNo,
    required this.defaultProbability,
    required this.isClosed,
    required this.isActive,
  });

  final String opportunityStageId;
  final String code;
  final String name;
  final int sequenceNo;
  final int defaultProbability;
  final bool isClosed;
  final bool isActive;

  factory OpportunityStage.fromJson(Map<String, dynamic> json) => OpportunityStage(
        opportunityStageId: json['opportunity_stage_id'] as String,
        code: json['code'] as String,
        name: json['name'] as String,
        sequenceNo: json['sequence_no'] as int? ?? 0,
        defaultProbability: json['default_probability'] as int? ?? 0,
        isClosed: json['is_closed'] as bool? ?? false,
        isActive: json['is_active'] as bool? ?? true,
      );
}

class CrmLookups {
  const CrmLookups({
    required this.pipelineStages,
    required this.opportunityStatuses,
    required this.leadStatuses,
    required this.activityTypes,
    required this.activityStatuses,
  });

  final List<String> pipelineStages;
  final List<String> opportunityStatuses;
  final List<String> leadStatuses;
  final List<String> activityTypes;
  final List<String> activityStatuses;

  static const defaults = CrmLookups(
    pipelineStages: [],
    opportunityStatuses: [],
    leadStatuses: [],
    activityTypes: ['NOTE', 'CALL', 'MEETING', 'EMAIL', 'TASK'],
    activityStatuses: [],
  );

  factory CrmLookups.fromJson(Map<String, dynamic> json) => CrmLookups(
        pipelineStages: (json['pipeline_stages'] as List<dynamic>? ?? [])
            .map((e) => e as String)
            .toList(),
        opportunityStatuses: (json['opportunity_statuses'] as List<dynamic>? ?? [])
            .map((e) => e as String)
            .toList(),
        leadStatuses: (json['lead_statuses'] as List<dynamic>? ?? [])
            .map((e) => e as String)
            .toList(),
        activityTypes: (json['activity_types'] as List<dynamic>? ?? [])
            .map((e) => e as String)
            .toList(),
        activityStatuses: (json['activity_statuses'] as List<dynamic>? ?? [])
            .map((e) => e as String)
            .toList(),
      );
}

class LookupService {
  LookupService({ApiClient? client}) : _client = client ?? ApiClient();

  final ApiClient _client;
  CrmLookups? _cache;

  void clearCache() => _cache = null;

  Future<CrmLookups> fetch() async {
    if (_cache != null) return _cache!;
    final res = await _client.get('/api/v1/crm/lookups');
    if (res.statusCode != 200) {
      return CrmLookups.defaults;
    }
    _cache = CrmLookups.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
    return _cache!;
  }

  Future<List<String>> pipelineStageCodes() async {
    final lookups = await fetch();
    return lookups.pipelineStages.isNotEmpty
        ? lookups.pipelineStages
        : defaultPipelineStages;
  }

  Future<List<String>> activityTypes() async {
    final lookups = await fetch();
    return lookups.activityTypes.isNotEmpty
        ? lookups.activityTypes
        : CrmLookups.defaults.activityTypes;
  }

  Future<List<ActivityType>> listActivityTypes({
    bool includeInactive = false,
  }) async {
    final res = await _client.get(
      '/api/v1/crm/activity-types',
      query: {'include_inactive': '$includeInactive'},
    );
    if (res.statusCode != 200) {
      throw Exception(_client.extractError(res.body) ?? 'Failed to load activity types');
    }
    final data = jsonDecode(res.body) as Map<String, dynamic>;
    return (data['items'] as List<dynamic>? ?? [])
        .cast<Map<String, dynamic>>()
        .map(ActivityType.fromJson)
        .toList();
  }

  Future<ActivityType> createActivityType({
    required String code,
    required String name,
  }) async {
    final res = await _client.post(
      '/api/v1/crm/activity-types',
      body: {'code': code, 'name': name},
    );
    if (res.statusCode != 201) {
      throw Exception(_client.extractError(res.body) ?? 'Failed to create activity type');
    }
    clearCache();
    return ActivityType.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  Future<ActivityType> updateActivityType(
    String activityTypeId, {
    bool? isActive,
    String? name,
  }) async {
    final body = <String, dynamic>{};
    if (isActive != null) body['is_active'] = isActive;
    if (name != null) body['name'] = name;
    final res = await _client.patch(
      '/api/v1/crm/activity-types/$activityTypeId',
      body: body,
    );
    if (res.statusCode != 200) {
      throw Exception(_client.extractError(res.body) ?? 'Failed to update activity type');
    }
    clearCache();
    return ActivityType.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  Future<List<ActivityOutcome>> listActivityOutcomes({
    String? activityTypeCode,
    bool includeInactive = false,
  }) async {
    final query = <String, String>{
      'include_inactive': '$includeInactive',
      if (activityTypeCode != null) 'activity_type_code': activityTypeCode,
    };
    final res = await _client.get('/api/v1/crm/activity-outcomes', query: query);
    if (res.statusCode != 200) {
      throw Exception(_client.extractError(res.body) ?? 'Failed to load outcomes');
    }
    final data = jsonDecode(res.body) as Map<String, dynamic>;
    return (data['items'] as List<dynamic>? ?? [])
        .cast<Map<String, dynamic>>()
        .map(ActivityOutcome.fromJson)
        .toList();
  }

  Future<ActivityOutcome> createActivityOutcome({
    required String activityTypeCode,
    required String code,
    required String name,
    bool isPositive = false,
  }) async {
    final res = await _client.post(
      '/api/v1/crm/activity-outcomes',
      body: {
        'activity_type_code': activityTypeCode,
        'code': code,
        'name': name,
        'is_positive': isPositive,
      },
    );
    if (res.statusCode != 201) {
      throw Exception(_client.extractError(res.body) ?? 'Failed to create outcome');
    }
    clearCache();
    return ActivityOutcome.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  Future<ActivityOutcome> updateActivityOutcome(
    String outcomeId, {
    bool? isActive,
    String? name,
    bool? isPositive,
  }) async {
    final body = <String, dynamic>{};
    if (isActive != null) body['is_active'] = isActive;
    if (name != null) body['name'] = name;
    if (isPositive != null) body['is_positive'] = isPositive;
    final res = await _client.patch(
      '/api/v1/crm/activity-outcomes/$outcomeId',
      body: body,
    );
    if (res.statusCode != 200) {
      throw Exception(_client.extractError(res.body) ?? 'Failed to update outcome');
    }
    clearCache();
    return ActivityOutcome.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  Future<List<OpportunityStage>> listOpportunityStages({
    bool includeInactive = true,
  }) async {
    final res = await _client.get(
      '/api/v1/crm/opportunity-stages',
      query: {'include_inactive': '$includeInactive'},
    );
    if (res.statusCode != 200) {
      return const [];
    }
    final data = jsonDecode(res.body) as Map<String, dynamic>;
    return (data['items'] as List<dynamic>? ?? [])
        .cast<Map<String, dynamic>>()
        .map(OpportunityStage.fromJson)
        .toList();
  }

  Future<OpportunityStage> createOpportunityStage({
    required String code,
    required String name,
    int defaultProbability = 0,
    bool isClosed = false,
  }) async {
    final res = await _client.post(
      '/api/v1/crm/opportunity-stages',
      body: {
        'code': code,
        'name': name,
        'default_probability': defaultProbability,
        'is_closed': isClosed,
      },
    );
    if (res.statusCode != 201) {
      throw Exception(_client.extractError(res.body) ?? 'Failed to create stage');
    }
    clearCache();
    return OpportunityStage.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  Future<OpportunityStage> updateOpportunityStage(
    String stageId, {
    bool? isActive,
    String? name,
  }) async {
    final body = <String, dynamic>{};
    if (isActive != null) body['is_active'] = isActive;
    if (name != null) body['name'] = name;
    final res = await _client.patch(
      '/api/v1/crm/opportunity-stages/$stageId',
      body: body,
    );
    if (res.statusCode != 200) {
      throw Exception(_client.extractError(res.body) ?? 'Failed to update stage');
    }
    clearCache();
    return OpportunityStage.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  Future<List<OpportunityStage>> reorderOpportunityStages(
    List<String> stageIds,
  ) async {
    final res = await _client.put(
      '/api/v1/crm/opportunity-stages/reorder',
      body: {'stage_ids': stageIds},
    );
    if (res.statusCode != 200) {
      throw Exception(_client.extractError(res.body) ?? 'Failed to reorder stages');
    }
    clearCache();
    final data = jsonDecode(res.body) as Map<String, dynamic>;
    return (data['items'] as List<dynamic>? ?? [])
        .cast<Map<String, dynamic>>()
        .map(OpportunityStage.fromJson)
        .toList();
  }
}

File: Backend/app/services/crm/lookup_service.py
from uuid import UUID, uuid4

from sqlalchemy.orm import Session

from app.core.deps import CurrentUser
from app.core.exceptions import AppError, ForbiddenError
from app.models.crm import ActivityType, ActivityOutcome, OpportunityStage
from app.repositories.crm.lookup_repository import LookupRepository
from app.schemas.crm.activity import ACTIVITY_STATUSES
from app.schemas.crm.lead import LEAD_STATUSES
from app.schemas.crm.lookup import (
    ActivityTypeCreate,
    ActivityTypeListResponse,
    ActivityTypeResponse,
    ActivityTypeUpdate,
    ActivityOutcomeListResponse,
    ActivityOutcomeResponse,
    ActivityOutcomeCreate,
    ActivityOutcomeUpdate,
    CrmLookupsResponse,
    OpportunityStageListResponse,
    OpportunityStageCreate,
    OpportunityStageReorder,
    OpportunityStageResponse,
    OpportunityStageUpdate,
)
from app.schemas.crm.opportunity import OPPORTUNITY_STATUSES, PIPELINE_STAGES, STAGE_PROBABILITY

DEFAULT_ACTIVITY_TYPES: tuple[tuple[str, str], ...] = (
    ("NOTE", "Note"),
    ("CALL", "Call"),
    ("MEETING", "Meeting"),
    ("EMAIL", "Email"),
    ("TASK", "Task"),
    ("SYSTEM", "System"),
)

DEFAULT_STAGES: tuple[tuple[str, str, int, int, bool], ...] = tuple(
    (
        code,
        code.replace("_", " ").title(),
        i + 1,
        STAGE_PROBABILITY.get(code, 0),
        False,
    )
    for i, code in enumerate(PIPELINE_STAGES)
)

DEFAULT_OUTCOMES: tuple[tuple[str, str, str, bool], ...] = (
    ("CALL", "INTERESTED", "Interested", True),
    ("CALL", "NOT_INTERESTED", "Not Interested", False),
    ("CALL", "CALLBACK", "Callback Requested", False),
    ("CALL", "NO_ANSWER", "No Answer", False),
    ("MEETING", "REQUIREMENTS_CAPTURED", "Requirements Captured", True),
    ("MEETING", "RESCHEDULED", "Rescheduled", False),
    ("MEETING", "CANCELLED", "Cancelled", False),
    ("TASK", "DONE", "Done", True),
    ("TASK", "DEFERRED", "Deferred", False),
)

OUTCOME_REQUIRED_TYPES = frozenset({"CALL", "MEETING", "TASK"})


class CrmLookupService:
    def __init__(self, db: Session) -> None:
        self.db = db
        self.repo = LookupRepository(db)

    def ensure_tenant_defaults(self, tenant_id: UUID) -> None:
        self.repo.seed_defaults(
            tenant_id, DEFAULT_ACTIVITY_TYPES, DEFAULT_STAGES, DEFAULT_OUTCOMES
        )

    def get_lookups(self, tenant_id: UUID) -> CrmLookupsResponse:
        self.ensure_tenant_defaults(tenant_id)
        activity_rows = self.repo.list_activity_types(tenant_id)
        stage_rows = self.repo.list_opportunity_stages(tenant_id)
        outcome_rows = self.repo.list_activity_outcomes(tenant_id)
        return CrmLookupsResponse(
            pipeline_stages=[s.code for s in stage_rows] or list(PIPELINE_STAGES),
            opportunity_statuses=list(OPPORTUNITY_STATUSES),
            lead_statuses=list(LEAD_STATUSES),
            activity_types=[a.code for a in activity_rows]
            or [c for c, _ in DEFAULT_ACTIVITY_TYPES],
            activity_statuses=list(ACTIVITY_STATUSES),
            activity_outcomes=[o.code for o in outcome_rows],
            activity_type_details=[
                ActivityTypeResponse.model_validate(a) for a in activity_rows
            ],
            activity_outcome_details=[
                ActivityOutcomeResponse.model_validate(o) for o in outcome_rows
            ],
            pipeline_stage_details=[
                OpportunityStageResponse.model_validate(s) for s in stage_rows
            ],
        )

    def list_activity_types(
        self, tenant_id: UUID, *, include_inactive: bool = False
    ) -> ActivityTypeListResponse:
        self.ensure_tenant_defaults(tenant_id)
        rows = self.repo.list_activity_types(
            tenant_id, active_only=not include_inactive
        )
        return ActivityTypeListResponse(
            items=[ActivityTypeResponse.model_validate(r) for r in rows]
        )

    def create_activity_type(
        self, tenant_id: UUID, current: CurrentUser, payload: ActivityTypeCreate
    ) -> ActivityTypeResponse:
        if current.role_code not in ("TENANT_ADMIN", "PLATFORM_ADMIN", "SALES_MANAGER"):
            raise ForbiddenError(
                "Activity type admin requires manager role", req_id="REQ-CRM-RBAC"
            )
        code = payload.code.strip().upper()
        if self.repo.get_activity_type_by_code(tenant_id, code):
            raise AppError(
                "VALIDATION_ERROR",
                f"Activity type '{code}' already exists",
                422,
                req_id="REQ-CRM-004",
            )
        row = ActivityType(
            activity_type_id=uuid4(),
            tenant_id=tenant_id,
            code=code,
            name=payload.name.strip(),
        )
        saved = self.repo.add_activity_type(row)
        return ActivityTypeResponse.model_validate(saved)

    def update_activity_type(
        self,
        tenant_id: UUID,
        current: CurrentUser,
        activity_type_id: UUID,
        payload: ActivityTypeUpdate,
    ) -> ActivityTypeResponse:
        if current.role_code not in ("TENANT_ADMIN", "PLATFORM_ADMIN", "SALES_MANAGER"):
            raise ForbiddenError(
                "Activity type admin requires manager role", req_id="REQ-CRM-RBAC"
            )
        row = self.repo.get_activity_type_by_id(tenant_id, activity_type_id)
        if row is None:
            raise AppError("NOT_FOUND", "Activity type not found", 404, req_id="REQ-CRM-004")
        data = payload.model_dump(exclude_unset=True)
        if "is_active" in data and data["is_active"] is not None:
            row.is_active = data["is_active"]
        if "name" in data and data["name"]:
            row.name = data["name"].strip()
        saved = self.repo.save_activity_type(row)
        return ActivityTypeResponse.model_validate(saved)

    def list_activity_outcomes(
        self,
        tenant_id: UUID,
        *,
        activity_type_code: str | None = None,
        include_inactive: bool = False,
    ) -> ActivityOutcomeListResponse:
        self.ensure_tenant_defaults(tenant_id)
        rows = self.repo.list_activity_outcomes(
            tenant_id,
            activity_type_code=activity_type_code,
            active_only=not include_inactive,
        )
        return ActivityOutcomeListResponse(
            items=[ActivityOutcomeResponse.model_validate(r) for r in rows]
        )

    def create_activity_outcome(
        self,
        tenant_id: UUID,
        current: CurrentUser,
        payload: ActivityOutcomeCreate,
    ) -> ActivityOutcomeResponse:
        if current.role_code not in ("TENANT_ADMIN", "PLATFORM_ADMIN", "SALES_MANAGER"):
            raise ForbiddenError(
                "Activity outcome admin requires manager role", req_id="REQ-CRM-RBAC"
            )
        type_code = payload.activity_type_code.strip().upper()
        code = payload.code.strip().upper()
        if self.repo.get_activity_outcome_by_code(tenant_id, type_code, code):
            raise AppError(
                "VALIDATION_ERROR",
                f"Outcome '{code}' already exists for {type_code}",
                422,
                req_id="REQ-CRM-004",
            )
        row = ActivityOutcome(
            activity_outcome_id=uuid4(),
            tenant_id=tenant_id,
            activity_type_code=type_code,
            code=code,
            name=payload.name.strip(),
            is_positive=payload.is_positive,
        )
        saved = self.repo.save_activity_outcome(row)
        return ActivityOutcomeResponse.model_validate(saved)

    def update_activity_outcome(
        self,
        tenant_id: UUID,
        current: CurrentUser,
        outcome_id: UUID,
        payload: ActivityOutcomeUpdate,
    ) -> ActivityOutcomeResponse:
        if current.role_code not in ("TENANT_ADMIN", "PLATFORM_ADMIN", "SALES_MANAGER"):
            raise ForbiddenError(
                "Activity outcome admin requires manager role", req_id="REQ-CRM-RBAC"
            )
        row = self.repo.get_activity_outcome_by_id(tenant_id, outcome_id)
        if row is None:
            raise AppError("NOT_FOUND", "Activity outcome not found", 404, req_id="REQ-CRM-004")
        data = payload.model_dump(exclude_unset=True)
        if "is_active" in data and data["is_active"] is not None:
            row.is_active = data["is_active"]
        if "name" in data and data["name"]:
            row.name = data["name"].strip()
        if "is_positive" in data and data["is_positive"] is not None:
            row.is_positive = data["is_positive"]
        saved = self.repo.save_activity_outcome(row)
        return ActivityOutcomeResponse.model_validate(saved)

    def list_opportunity_stages(
        self, tenant_id: UUID, *, include_inactive: bool = False
    ) -> OpportunityStageListResponse:
        self.ensure_tenant_defaults(tenant_id)
        rows = self.repo.list_opportunity_stages(
            tenant_id, active_only=not include_inactive
        )
        return OpportunityStageListResponse(
            items=[OpportunityStageResponse.model_validate(r) for r in rows]
        )

    def create_opportunity_stage(
        self,
        tenant_id: UUID,
        current: CurrentUser,
        payload: OpportunityStageCreate,
    ) -> OpportunityStageResponse:
        if current.role_code not in ("TENANT_ADMIN", "PLATFORM_ADMIN", "SALES_MANAGER"):
            raise ForbiddenError(
                "Stage admin requires manager role", req_id="REQ-CRM-RBAC"
            )
        self.ensure_tenant_defaults(tenant_id)
        code = payload.code.strip().upper()
        if self.repo.get_opportunity_stage_by_code(tenant_id, code):
            raise AppError(
                "VALIDATION_ERROR",
                f"Pipeline stage '{code}' already exists",
                422,
                req_id="REQ-CRM-002",
            )
        existing = self.repo.list_opportunity_stages(tenant_id, active_only=False)
        next_seq = max((s.sequence_no for s in existing), default=0) + 1
        row = OpportunityStage(
            opportunity_stage_id=uuid4(),
            tenant_id=tenant_id,
            code=code,
            name=payload.name.strip(),
            sequence_no=next_seq,
            default_probability=payload.default_probability,
            is_closed=payload.is_closed,
        )
        saved = self.repo.save_opportunity_stage(row)
        return OpportunityStageResponse.model_validate(saved)

    def update_opportunity_stage(
        self,
        tenant_id: UUID,
        current: CurrentUser,
        stage_id: UUID,
        payload: OpportunityStageUpdate,
    ) -> OpportunityStageResponse:
        if current.role_code not in ("TENANT_ADMIN", "PLATFORM_ADMIN", "SALES_MANAGER"):
            raise ForbiddenError(
                "Stage admin requires manager role", req_id="REQ-CRM-RBAC"
            )
        row = self.repo.get_opportunity_stage_by_id(tenant_id, stage_id)
        if row is None:
            raise AppError("NOT_FOUND", "Opportunity stage not found", 404, req_id="REQ-CRM-002")
        data = payload.model_dump(exclude_unset=True)
        if "is_active" in data and data["is_active"] is not None:
            row.is_active = data["is_active"]
        if "name" in data and data["name"]:
            row.name = data["name"].strip()
        if "sequence_no" in data and data["sequence_no"] is not None:
            row.sequence_no = data["sequence_no"]
        saved = self.repo.save_opportunity_stage(row)
        return OpportunityStageResponse.model_validate(saved)

    def reorder_opportunity_stages(
        self,
        tenant_id: UUID,
        current: CurrentUser,
        payload: OpportunityStageReorder,
    ) -> OpportunityStageListResponse:
        if current.role_code not in ("TENANT_ADMIN", "PLATFORM_ADMIN", "SALES_MANAGER"):
            raise ForbiddenError(
                "Stage admin requires manager role", req_id="REQ-CRM-RBAC"
            )
        self.ensure_tenant_defaults(tenant_id)
        existing = self.repo.list_opportunity_stages(tenant_id, active_only=False)
        existing_ids = {s.opportunity_stage_id for s in existing}
        if set(payload.stage_ids) != existing_ids:
            raise AppError(
                "VALIDATION_ERROR",
                "Reorder must include every tenant stage exactly once",
                422,
                req_id="REQ-CRM-002",
            )
        try:
            rows = self.repo.reorder_opportunity_stages(tenant_id, payload.stage_ids)
        except ValueError as exc:
            raise AppError("VALIDATION_ERROR", str(exc), 422, req_id="REQ-CRM-002") from exc
        active = [r for r in rows if r.is_active]
        return OpportunityStageListResponse(
            items=[OpportunityStageResponse.model_validate(r) for r in active]
        )
